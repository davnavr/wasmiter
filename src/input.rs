use std::io::Read;

/// An extension of the [`Read`] trait that provides support for creating new readers over the same
/// input at a particular location.
pub trait Input: Read {
    /// A reader used to read a portion of the [`Input`] at a particular location.
    type Portion: Read;

    /// Creates a new reader that reads byte starting at the current location.
    fn portion(&mut self) -> Self::Portion;
}

impl<'a> Input for &'a [u8] {
    type Portion = &'a [u8];

    fn portion(&mut self) -> Self::Portion {
        self
    }
}
