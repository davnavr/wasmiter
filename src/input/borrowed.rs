/// Trait for creating a clone of an [`Input`](crate::input::Input) by borrowing it.
pub trait BorrowInput {
    /// The type of the clone with the borrowed [`Input`](crate::input::Input).
    type Borrowed<'a> where Self: 'a;

    /// Borrows the underlying [`Input`](crate::input::Input).
    fn borrow_input(&self) -> Self::Borrowed<'_>;
}
