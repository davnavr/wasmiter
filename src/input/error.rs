use core::fmt::{Display, Formatter};

#[derive(Clone, Copy, Debug)]
pub(super) enum ErrorKind {
    OutOfBounds,
    CannotFillBuffer,
    OffsetOverflow,
}

impl Display for ErrorKind {
    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
        f.write_str(match self {
            Self::OutOfBounds => "operation would be out of bounds",
            Self::CannotFillBuffer => "buffer could not be completely filled",
            Self::OffsetOverflow => "offset would overflow",
        })
    }
}

cfg_if::cfg_if! {
    if #[cfg(feature = "alloc")] {
        struct BoxedError {
            kind: ErrorKind,
            offset: u64,
            length: Option<u64>
        }

        type ErrorInner = alloc::boxed::Box<BoxedError>;
    } else {
        enum InlineError {
            NoOffset(ErrorKind),
            WithOffset {
                kind: ErrorKind,
                offset: u32,
            },
        }

        type ErrorInner = InlineError;
    }
}

/// Error type used when an operation to read [`Input`](crate::input::Input) fails.
#[repr(transparent)]
pub struct Error {
    inner: ErrorInner,
}

impl Error {
    #[must_use]
    pub(super) fn new(kind: ErrorKind, offset: u64, length: Option<u64>) -> Self {
        cfg_if::cfg_if! {
            if #[cfg(feature = "alloc")] {
                Self {
                    inner: alloc::boxed::Box::new(BoxedError { kind, offset, length }),
                }
            } else {
                let _ = length;
                Self {
                    inner: if let Ok(small_offset) = u32::try_from(offset) {
                        InlineError::WithOffset { kind, offset: small_offset }
                    } else {
                        InlineError::NoOffset(kind)
                    },
                }
            }
        }
    }

    fn kind(&self) -> ErrorKind {
        cfg_if::cfg_if! {
            if #[cfg(feature = "alloc")] {
                self.inner.kind
            } else {
                let (InlineError::WithOffset { kind, .. } | InlineError::NoOffset(kind)) = self.inner;
                kind
            }
        }
    }
}

#[cfg(feature = "alloc")]
impl Error {
    /// Gets the offset from the start of the [`Input`](crate::input::Input) where the error
    /// occured.
    #[inline]
    pub fn offset(&self) -> u64 {
        self.inner.offset
    }

    /// Gets the remaining length of the [`Input`](crate::input::Input) where the error occured.
    #[inline]
    pub fn length(&self) -> Option<u64> {
        self.inner.length
    }
}

impl core::fmt::Debug for Error {
    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
        let mut s = f.debug_struct("Error");
        s.field("kind", &self.kind());

        cfg_if::cfg_if! {
            if #[cfg(feature = "alloc")] {
                s.field("offset", &self.offset()).field("length", &self.length());
            } else {
                if let InlineError::WithOffset { offset, .. } = self.inner {
                    s.field("offset", &offset);
                }
            }
        }

        s.finish()
    }
}

impl Display for Error {
    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
        write!(f, "{}", self.kind())?;

        cfg_if::cfg_if! {
            if #[cfg(feature = "alloc")] {
                write!(f, " at offset {}", self.offset())?;

                if let Some(length) = self.length() {
                    write!(f, ", length of input is {}", length)?;
                }
            } else {
                if let InlineError::WithOffset { offset, .. } = self.inner {
                    write!(f, " at offset {}", offset)?;
                }
            }
        }

        Ok(())
    }
}

#[cfg(feature = "std")]
impl std::error::Error for Error {}
