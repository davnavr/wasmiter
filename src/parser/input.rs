use std::fs::File;
use std::io::{Read, Result, Seek, SeekFrom};

/// Allows reading bytes from some source.
///
/// [`Input`] implementations are expected to keep track of their position within the
/// original source.
pub trait Input: Sized {
    /// The [`Read`] implementation used to read bytes.
    ///
    /// Calls to read bytes from the input **must** advance the [`Input`]'s position.
    type Reader<'a>: Read
    where
        Self: 'a;

    /// Returns a duplicate of the [`Input`] at the current position.
    ///
    /// Using the [`Reader`](Input::Reader) of the duplicate **does not** change the position of
    /// the original.
    fn fork(&self) -> Result<Self>;

    /// Returns a reader used to read bytes.
    ///
    /// Consuming bytes from this reader advances the [`position`](Input::position).
    fn reader(&mut self) -> Result<Self::Reader<'_>>;

    /// Gets current position, which is a byte offset from the start of the input.
    fn position(&self) -> Result<u64>;
}

impl<'a> Input for std::io::Cursor<&'a [u8]> {
    type Reader<'b> = &'b mut Self where 'a: 'b;

    #[inline]
    fn fork(&self) -> Result<Self> {
        Ok(self.clone())
    }

    #[inline]
    fn reader(&mut self) -> Result<Self::Reader<'_>> {
        Ok(self)
    }

    #[inline]
    fn position(&self) -> Result<u64> {
        Ok(self.position())
    }
}

/// Reads bytes from a [`File`] starting at a given location.
#[derive(Debug)]
pub struct FileReader<'a> {
    offset: &'a mut Result<u64>,
    file: &'a File,
}

impl Read for FileReader<'_> {
    #[inline]
    fn read(&mut self, buf: &mut [u8]) -> Result<usize> {
        self.file.read(buf)
    }

    #[inline]
    fn read_exact(&mut self, buf: &mut [u8]) -> Result<()> {
        self.file.read_exact(buf)
    }
}

impl Drop for FileReader<'_> {
    fn drop(&mut self) {
        *self.offset = self.file.stream_position();
    }
}

/// Allows reading bytes from different parts of a [`File`].
#[derive(Debug)]
pub struct FileInput {
    offset: Result<u64>,
    file: File,
}

impl FileInput {
    /// Opens the file at the given [`Path`](std::path::Path) as [`Input`].
    pub fn from_path<P: AsRef<std::path::Path>>(path: P) -> Result<Self> {
        Ok(Self {
            offset: Ok(0),
            file: File::open(path)?,
        })
    }
}

impl Input for FileInput {
    type Reader<'b> = FileReader<'b> where Self: 'b;

    fn position(&self) -> Result<u64> {
        #[inline(never)]
        #[cold]
        fn clone_error(error: &std::io::Error) -> std::io::Error {
            std::io::Error::new(
                error.kind(),
                format!("unable to access underlying reader: {error}"),
            )
        }

        match self.offset.as_ref() {
            Ok(offset) => Ok(*offset),
            Err(e) => Err(clone_error(e)),
        }
    }

    fn fork(&self) -> Result<Self> {
        Ok(FileInput {
            offset: Ok(self.position()?),
            file: self.file.try_clone()?,
        })
    }

    fn reader(&mut self) -> Result<FileReader<'_>> {
        self.file.seek(SeekFrom::Start(self.position()?))?;
        Ok(FileReader {
            offset: &mut self.offset,
            file: &self.file,
        })
    }
}
