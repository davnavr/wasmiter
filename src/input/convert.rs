use crate::input::Input;

/// Trait for types that have an [`Input`].
pub trait HasInput<I: Input> {
    /// Gets the underlying [`Input`].
    fn input(&self) -> &I;
}

/// Trait for creating a clone by borrowing an underlying [`Input`].
pub trait BorrowInput<'a, I: Input + 'a>: HasInput<I> {
    /// The type with the borrowed [`Input`].
    type Borrowed: HasInput<&'a I>;

    /// Creates a clone by borrows the underlying [`Input`].
    fn borrow_input(&'a self) -> Self::Borrowed;
}

/// Trait for cloning an underlying [`Input`].
pub trait CloneInput<'a, I: Clone + Input + 'a>: HasInput<&'a I> {
    /// The type with the cloned [`Input`].
    type Cloned: HasInput<I>;

    /// Clones the underlying [`Input`].
    fn clone_input(&self) -> Self::Cloned;
}
