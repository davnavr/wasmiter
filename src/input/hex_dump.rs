use crate::input::{BorrowInput, Input, Result, Window};
use core::{
    fmt::{Debug, Display, Formatter, LowerHex, UpperHex, Write as _},
    num::NonZeroU8,
};

/// A row of at most 16 bytes obtained from a [`HexDump`].
#[derive(Clone, Copy)]
pub struct HexDumpRow {
    offset: u64,
    count: NonZeroU8,
    bytes: [u8; 16],
}

impl HexDumpRow {
    /// The offset of into the [`Input`] to the first byte of the [`Row`]'s contents.
    #[inline]
    pub fn offset(&self) -> u64 {
        self.offset
    }

    /// Gets the contents of the row as a slice.
    #[inline]
    pub fn contents(&self) -> &[u8] {
        &self.bytes[..self.count.get().into()]
    }

    fn bytes(&self) -> impl core::iter::FusedIterator<Item = (u64, u8)> + '_ {
        (self.offset..=u64::MAX).zip(self.contents().iter().copied())
    }

    #[inline]
    fn fmt_hex(
        &self,
        f: &mut Formatter<'_>,
        mut hex: impl FnMut(u8, &mut Formatter<'_>) -> core::fmt::Result,
    ) -> core::fmt::Result {
        if f.alternate() {
            write!(f, "0x")?;
        }

        let mut first = true;
        for (offset, byte) in self.bytes() {
            if !first {
                if offset % 4 == 0 {
                    f.write_char('_')?;
                }

                if offset % 8 == 0 {
                    f.write_char('_')?;
                }
            }

            first = false;
            hex(byte, f)?;
        }

        Ok(())
    }

    fn fmt_display(&self, f: &mut Formatter<'_>, offset_width: usize) -> core::fmt::Result {
        if f.alternate() {
            write!(f, "{:#offset_width$X}  ", self.offset)?;
        }

        // Padding
        {
            let pad_amount = (self.offset % 16) as u8 * 3 + u8::from(self.offset >= 8);
            for _ in 0..pad_amount {
                f.write_char(' ')?;
            }
        }

        // Bytes
        let mut first = true;
        for (offset, byte) in self.bytes() {
            if !first {
                f.write_char(' ')?;

                if offset % 16 == 8 {
                    f.write_char(' ')?;
                }
            }

            first = false;
            write!(f, "{byte:X}")?;
        }

        if f.alternate() {
            f.write_str("  |")?;

            for _ in 0..(self.offset % 16) {
                f.write_char(' ')?;
            }

            for byte in self.contents().iter() {
                if matches!(byte, 0x20..=0x7E) {
                    if let Some(c) = char::from_u32(u32::from(*byte)) {
                        f.write_char(c)?;
                    } else {
                        return Err(core::fmt::Error);
                    }
                } else {
                    f.write_char('.')?;
                }
            }

            f.write_char('|')?;
        }

        Ok(())
    }
}

impl UpperHex for HexDumpRow {
    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
        self.fmt_hex(f, |byte, f| write!(f, "{byte:X}"))
    }
}

impl LowerHex for HexDumpRow {
    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
        self.fmt_hex(f, |byte, f| write!(f, "{byte:x}"))
    }
}

impl Debug for HexDumpRow {
    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
        #[repr(transparent)]
        struct Contents<'a>(&'a HexDumpRow);

        impl Debug for Contents<'_> {
            fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
                write!(f, "{:X}", self.0)
            }
        }

        if !f.alternate() {
            f.debug_struct("Row")
                .field("offset", &self.offset)
                .field("contents", &Contents(self))
                .finish()
        } else {
            UpperHex::fmt(&self, f)
        }
    }
}

impl Display for HexDumpRow {
    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
        self.fmt_display(f, 4)
    }
}

/// Iterates over the contents of an [`Input`], also providing a [`Debug`](core::fmt::Debug) and
/// [`Display`] implementation.
pub struct HexDump<I: Input> {
    window: Window<I>,
}

impl<I: Input> From<Window<I>> for HexDump<I> {
    #[inline]
    fn from(window: Window<I>) -> Self {
        Self { window }
    }
}

impl<I: Input> BorrowInput for HexDump<I> {
    type Borrowed<'a> = HexDump<&'a I> where I: 'a;

    fn borrow_input(&self) -> HexDump<&I> {
        HexDump {
            window: self.window.borrow_input(),
        }
    }
}

impl<I: Input> Iterator for HexDump<I> {
    type Item = Result<HexDumpRow>;

    fn next(&mut self) -> Option<Self::Item> {
        if self.window.length() == 0 {
            return None;
        }

        let mut buffer = [0u8; 16];
        let start = self.window.base();
        let result = self.window.read_at(start, &mut buffer).and_then(|buf| {
            let len = buf.len() as u8;
            self.window.advance(len.into())?;
            Ok(NonZeroU8::new(len))
        });

        let length = match result {
            Ok(Some(len)) => Ok(len),
            Ok(None) => Err(None),
            Err(e) => Err(Some(e)),
        };

        match length {
            Ok(count) => Some(Ok(HexDumpRow {
                offset: start,
                count,
                bytes: buffer,
            })),
            Err(e) => {
                self.window.shrink(self.window.length());
                e.map(Err)
            }
        }
    }
}

impl<I: Input> Display for HexDump<I> {
    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
        const OFFSET_HEADER: &str = "offset";

        let max_width = self
            .window
            .base()
            .checked_add(self.window.length())
            .and_then(|max| {
                usize::try_from(max.checked_ilog(16).unwrap_or(0u32))
                    .ok()?
                    .checked_add(1)
            })
            .unwrap_or(4);
        let width = core::cmp::max(max_width, OFFSET_HEADER.len());

        if f.alternate() {
            write!(f, "{:<width$}  ", OFFSET_HEADER)?;
        }

        writeln!(f, " 0  1  2  3  4  5  6  7   8  9  A  B  C  D  E  F")?;

        for row in self.borrow_input().filter_map(Result::ok) {
            Display::fmt(&row, f)?;
            writeln!(f)?;
        }

        Ok(())
    }
}

impl<I: Input> Debug for HexDump<I> {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        f.debug_list().entries(self.borrow_input()).finish()
    }
}
