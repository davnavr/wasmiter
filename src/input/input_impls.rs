use crate::input::{self, Input, Result};

impl Input for [u8] {
    #[inline]
    fn read_at<'b>(&self, offset: u64, buffer: &'b mut [u8]) -> Result<&'b mut [u8]> {
        let start = usize::try_from(offset).unwrap_or(usize::MAX);
        if let Some(source) = self.get(start..) {
            let copy_amount = core::cmp::min(source.len(), buffer.len());
            let destination = &mut buffer[..copy_amount];
            destination.copy_from_slice(&source[..copy_amount]);
            Ok(destination)
        } else {
            Err(input::out_of_bounds(
                offset,
                self.len()
                    .checked_sub(start)
                    .and_then(|len| u64::try_from(len).ok()),
            ))
        }
    }

    #[inline]
    fn length_at(&self, offset: u64) -> Result<u64> {
        let length = usize::try_from(offset)
            .ok()
            .and_then(|start| self.len().checked_sub(start))
            .and_then(|len| u64::try_from(len).ok());

        if let Some(len) = length {
            Ok(len)
        } else {
            Err(input::out_of_bounds(offset, None))
        }
    }

    #[inline]
    fn try_eq_at(&self, offset: u64, bytes: &[u8]) -> Result<bool> {
        if let Some(input) = usize::try_from(offset)
            .ok()
            .and_then(|start| self.get(start..))
        {
            Ok(&input[..input.len().min(bytes.len())] == bytes)
        } else {
            Err(input::out_of_bounds(offset, None))
        }
    }
}

/// Allows reading bytes from a memory map.
///
/// # Safety
///
/// Although UB is easy if underlying storage is modified, **pretending** it doesn't happen should
/// be fine, since bytes will still be read no matter what (their values don't matter).
#[cfg(feature = "mmap")]
impl Input for memmap2::Mmap {
    #[inline]
    fn read_at<'b>(&self, offset: u64, buffer: &'b mut [u8]) -> Result<&'b mut [u8]> {
        <[u8] as Input>::read_at(self.as_ref(), offset, buffer)
    }

    #[inline]
    fn length_at(&self, offset: u64) -> Result<u64> {
        <[u8] as Input>::length_at(self, offset)
    }

    #[inline]
    fn try_eq_at(&self, offset: u64, bytes: &[u8]) -> Result<bool> {
        <[u8] as Input>::try_eq_at(self, offset, bytes)
    }
}

macro_rules! delegated_input_impl {
    ($b:ident in $($implementor:ty $(,)?)+) => {$(
        impl<$b: Input + ?Sized> Input for $implementor {
            #[inline]
            fn read_at<'b>(&self, offset: u64, buffer: &'b mut [u8]) -> Result<&'b mut [u8]> {
                <$b as Input>::read_at(self, offset, buffer)
            }

            #[inline]
            fn length_at(&self, offset: u64) -> Result<u64> {
                <$b as Input>::length_at(self, offset)
            }

            #[inline]
            fn try_eq_at(&self, offset: u64, bytes: &[u8]) -> Result<bool> {
                <$b as Input>::try_eq_at(self, offset, bytes)
            }

            #[inline]
            fn read_exact_at(&self, offset: u64, buffer: &mut [u8]) -> Result<()> {
                <$b as Input>::read_exact_at(self, offset, buffer)
            }

            #[inline]
            fn read<'b>(&self, offset: &mut u64, buffer: &'b mut [u8]) -> Result<&'b mut [u8]> {
                <$b as Input>::read(self, offset, buffer)
            }

            #[inline]
            fn read_exact(&self, offset: &mut u64, buffer: &mut [u8]) -> Result<()> {
                <$b as Input>::read_exact(self, offset, buffer)
            }
        }
    )*};
}

delegated_input_impl!(I in &I);

#[cfg(feature = "alloc")]
delegated_input_impl!(I in alloc::sync::Arc<I>, alloc::rc::Rc<I>, alloc::boxed::Box<I>);
