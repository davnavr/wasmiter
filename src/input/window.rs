use crate::input::{self, BorrowInput, Input, Result};

/// Adapts an [`Input`] implementation to limit the amount of bytes that can be read to a specific
/// range.
///
/// # Example
///
/// ```
/// use wasmiter::input::{Input, Window};
///
/// let bytes: &[u8] = b"This is a test of the Window struct";
/// let window = Window::with_offset_and_length(bytes, 10, 18); // "test of the Window"
///
/// assert_eq!(window.length_at(22)?, 6);
///
/// let mut buffer = [0u8; 6];
/// let copied = window.read_at(15, &mut buffer)?;
/// assert_eq!(copied, b"of the");
/// # wasmiter::input::Result::Ok(())
/// ```
#[derive(Clone, Copy)]
pub struct Window<I: Input> {
    base: u64,
    length: u64,
    inner: I,
}

impl<I: Input> Window<I> {
    /// Creates a new [`Window`] into the specified [`Input`] that ensures reads can only occur at
    /// the given `offset` for `length` bytes.
    pub fn with_offset_and_length(inner: I, offset: u64, length: u64) -> Self {
        Self {
            base: offset,
            length,
            inner,
        }
    }

    /// Creates a new [`Window`] into the specified [`Input`] that ensures reads can only occur
    /// starting at the given `offset`.
    ///
    /// The [`length`](Window::length) is set to [`u64::MAX`].
    #[inline]
    pub fn with_offset(inner: I, offset: u64) -> Self {
        Self::with_offset_and_length(inner, offset, u64::MAX)
    }

    /// Gets the offset at which the [`Window`]'s content begins.
    #[inline]
    pub fn base(&self) -> u64 {
        self.base
    }

    /// Gets length, in bytes, of the [`Window`].
    #[inline]
    pub fn length(&self) -> u64 {
        self.length
    }

    /// Gets a reference to the inner [`Input`], bypassing the [`Window`]'s restrictions.
    #[inline]
    pub fn as_inner(&self) -> &I {
        &self.inner
    }

    /// Returns the inner [`Input`], bypassing the [`Window`]'s restrictions.
    #[inline]
    pub fn into_inner(self) -> I {
        self.inner
    }

    #[inline]
    pub(super) fn advance(&mut self, amount: u64) -> Result<()> {
        if self.length > 0 {
            input::increment_offset(&mut self.base, amount)?;
            self.shrink(amount);
            Ok(())
        } else {
            Err(input::offset_overflowed(self.base))
        }
    }

    /// Reduces the [`length`](Window::length) of the [`Window`] by the given `amount`, returning
    /// the new length.
    #[inline]
    pub(super) fn shrink(&mut self, amount: u64) {
        self.length = self.length.saturating_sub(amount);
    }

    #[inline]
    fn bounds_check(&self, offset: u64) -> Result<u64> {
        let (end, overflow) = self.base.overflowing_add(self.length);
        if offset >= self.base && offset < end && !overflow {
            Ok(self.length - (offset - self.base))
        } else {
            Err(input::Error::new(
                input::error::ErrorKind::OutOfBounds,
                offset,
                self.inner.length_at(offset).ok(),
            ))
        }
    }
}

impl<I: Input> From<I> for Window<I> {
    /// Creates a new [`Window`] into the specified [`Input`] that allows reading the first
    /// [`u64::MAX`] bytes.
    #[inline]
    fn from(inner: I) -> Self {
        Self::with_offset(inner, 0)
    }
}

impl<I: Input> Input for Window<I> {
    fn read_at<'b>(&self, offset: u64, buffer: &'b mut [u8]) -> Result<&'b mut [u8]> {
        let remaining = usize::try_from(self.bounds_check(offset)?).unwrap_or(usize::MAX);
        let actual_len = core::cmp::min(remaining, buffer.len());
        self.inner.read_at(offset, &mut buffer[0..actual_len])
    }

    fn length_at(&self, offset: u64) -> Result<u64> {
        Ok(core::cmp::min(
            self.bounds_check(offset)?,
            self.inner.length_at(offset)?,
        ))
    }
}

impl<I: Input> BorrowInput for Window<I> {
    type Borrowed<'a> = Window<&'a I> where I: 'a;

    fn borrow_input(&self) -> Self::Borrowed<'_> {
        Window {
            base: self.base,
            length: self.length,
            inner: &self.inner,
        }
    }
}

impl<I: Input> core::fmt::Debug for Window<I> {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        core::fmt::Debug::fmt(&input::HexDump::from(self.borrow_input()), f)
    }
}
