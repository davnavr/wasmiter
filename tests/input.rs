use wasmiter::input::{HexDump, Window};

const DATA: &[u8] =
    b"and the man went out for jury duty and exclaimed, \"Segmentation fault (core dumped)\"\x00";

#[test]
fn hex_dump_debug() {
    insta::assert_debug_snapshot!(HexDump::from(DATA));
}

#[test]
fn hex_dump_display() {
    insta::assert_display_snapshot!(format_args!("{:#}", HexDump::from(DATA)));
}

#[test]
fn hex_dump_display_at_weird_offset() {
    let window = Window::with_offset_and_length(DATA, 3, 45);
    insta::assert_display_snapshot!(format_args!("{:#}", HexDump::from(window)));
}
