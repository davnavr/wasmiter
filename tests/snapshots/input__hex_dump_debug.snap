---
source: tests/input.rs
expression: "HexDump::from(DATA)"
---
[
    Ok(
        0x616E6420_74686520__6D616E20_77656E74,
    ),
    Ok(
        0x206F7574_20666F72__206A7572_79206475,
    ),
    Ok(
        0x74792061_6E642065__78636C61_696D6564,
    ),
    Ok(
        0x2C202253_65676D65__6E746174_696F6E20,
    ),
    Ok(
        0x6661756C_74202863__6F726520_64756D70,
    ),
    Ok(
        0x65642922_0,
    ),
]
